### PR TITLE
Add Knowledge Checks with links and revise Learning Outcomes in Rails Advanced Forms lesson

### LINKED ISSUE #22948: Add missing Knowledge Checks, Learning Outcomes, and Knowledge Check links to Ruby on Rails/Advanced Forms
Related to #22470

This task is to go through the Advanced Forms lesson in the Ruby on Rails course and (as applicable):

- Add missing Learning Outcomes
- Add missing Knowledge Checks
- Add answer links in the Knowledge Checks

If the lesson already has Learning Outcomes or Knowledge Checks they do not need to be added to. Many lessons only need answer links added.

Knowledge check links should consist of an `<a>` tag which adds the `knowledge-check-link` class and links to the part of the lesson (or the external resource that is linked in the lesson) where the answer to the Knowledge Check can be found. Knowledge Check links should **not** point to external resources that weren't linked to as part of the lesson content or assignment. You can link to a specific line in the lesson like so:

```html
Lorem ipsum dolor sit amet, <span id='example-knowledge-check'>consectetur adipiscing elit</span>, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

...
<!-- This a tag just goes to the page where the span with this id lives -->
<a class='knowledge-check-link' href='#example-knowledge-check'>How do you return a value from a function?</a>
```
Or you can link to just a section of the lesson by just linking to the section heading and following [these](https://github.com/TheOdinProject/curriculum/wiki/Contributing-to-Wiki#anchor-linking) rules for determining how to format the link. Example:

```html
### Lorem ipsum dolor sit amet

sed do eiusmod tempor incididunt ut labore et dolore magna aliqua

...
<!-- This a tag just goes to a heading on the page-->
<a class='knowledge-check-link' href='#lorem-ipsum-dolor-sit-amet'>How do you return a value from a function?</a>
```

### Acceptance Criteria

- [ ] Lesson referenced in the title has Learning Outcomes, Knowledge Checks, and Knowledge Check links
- [ ] Knowledge Check links all link to the proper place in the lesson, or the proper external resource that was linked to earlier in the lesson
- [ ] Knowledge Check styling should **not** change

Please do not start work on this unless you've been assigned to it. Comment on this issue and ping me @wise-king-sullyman if you would like to be assigned to it.

Additionally, we would like for contributors to only take on one of these issues at a time. Please only request assignment for (and begin work on) an issue after the Learning Outcome/Knowledge Check/Knowledge Check link issue you're currently assigned to has been approved and merged.

Please check out [this new post on our wiki](https://github.com/TheOdinProject/curriculum/wiki/Learning-Outcomes-vs-Knowledge-Checks) about Learning Outcomes and Knowledge Checks if you need to write new Learning Outcomes or Knowledge Checks for this issue.

### PR DESCRIPTION
**Pull request by dm-murphy**
- [X] You have read [The Odin Projects Contributing Guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md).
 - [X] The title summarizes the change and where it happened, for example: "Fixes punctuation in Clean Code lesson".
 - [X] If the PR is related to an open issue, use a [relevant keyword](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) and reference it with the `#` sign and the issue number.
 - [X] You have previewed your markdown (if any) in the [Odin Markdown Preview Tool](https://www.theodinproject.com/lessons/preview)
 - [X] If changes are requested, please make the changes in a timely manner. After you submit the changes, request another review from the maintainer (top right).

#### 1. Describe the changes made and include why they are necessary or important:

This PR adds a Knowledge Checks section to the Rails Advanced Forms lesson. The Knowledge Checks are reworked from the original Learning Outcomes questions. The Knowledge Checks link back to answers via section headers, specific lines and one assignment resource. This PR also adds new Learning Outcomes bullets at a high-level of the lesson. Please let me know any revisions you would like on these, thanks!

Please note: While working on this lesson I detected an issue with one of the section headers on the live site. The header marked "Prepopulating select Tags with Collections" has an extraneous title in the h3 tag prior to the anchor link. It only appears on the live site and does not appear in the markdown file. It looks like one of these section headers was meant to replace the other, but both show up on the live site. I just wanted to point that out, thanks!

#### 2. Related Issue

Closes #22948

### CODE REVIEW

**Comment by wise-king-sullyman** [comment]
I actually need to go through this lesson before I can give a final review, but from what I can comment on so far I think that the learning outcomes aren't quite as high level as I would like to see for the most part. `* Make nested forms` is solid, `* Whitelist nested parameters` seems good as well, but the rest feel a bit too specific to me.

**Comment by dm-murphy** [comment]
> I actually need to go through this lesson before I can give a final review (if any maintainers that have already done this lesson would like to review it please do so), but from what I can comment on so far I think that the learning outcomes aren't quite as high level as I would like to see for the most part. `* Make nested forms` is solid, `* Whitelist nested parameters` seems good as well, but the rest feel a bit too specific to me.

Thanks Austin, appreciate the feedback. I can sit on this for a few days to give you or another maintainer a chance to make any suggestions. Otherwise I will redraft the other bullet points and submit for re-review. Thanks again, Daniel

File: rails_programming/advanced_forms_and_activerecord/forms_advanced.md [hunk 1]
```diff
@@ -193,3 +193,14 @@ This section contains helpful links to other content. It isn't required, so cons
...
+
+- <a class="knowledge-check-link" href='#prepopulating-select-tags-with-collections'>What does the `#select_tag` helper do?</a>
+- <a class="knowledge-check-link" href='#select-knowledge-check'>When would you use the `#select` helper?</a>
+- <a class="knowledge-check-link" href='#options-knowledge-check'>When using `#options_for_select`, what format does the array need to be in?</a>
+- <a class="knowledge-check-link" href='#nested-forms'>How can you prevent users from having to submit multiple forms?</a>
+- <a class="knowledge-check-link" href='#model-knowledge-check'>What do you add to the model that allows nested forms to create new objects?</a>
+- <a class="knowledge-check-link" href='https://www.createdbypete.com/2014/04/04/working-with-nested-forms-and-a-many-to-many-association-in-rails-4.html'>How do you whitelist the nested parameters in your controller?</a>
+ - <a class="knowledge-check-link" href='#miscellania-blank-submissions-that-mean-delete'>What happens if you leave a form field (e.g. a checkbox) blank (unchecked)?</a>
```
Comment (wise-king-sullyman) [inline rails_programming/advanced_forms_and_activerecord/forms_advanced.md:206]: I think this question could be improved with a slight reworking. Currently, it implies that there is an absolute that occurs when a field is left blank when in reality it varies.

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
I have a few additional comments after reviewing the lesson.

Also, I don't think that including `knowledge-check` in the ID names is needed, i.e. I think that `id='options-for-select'`  would be preferable to `id='options-knowledge-check'`

File: rails_programming/advanced_forms_and_activerecord/forms_advanced.md (hunk 1 above)
Comment (dm-murphy) [reply rails_programming/advanced_forms_and_activerecord/forms_advanced.md:206]: I see what you mean. I'll rework this one, thanks!

**Comment by dm-murphy** [comment]
> 

Gotcha! Thanks, I'll fix up all the ID names too along with the rest of the changes and resubmit for review.

**Review by wise-king-sullyman — APPROVED** [review]
These look great! Thank you for making these changes/this contribution in general!

**Pull request merged**